### ISSUE TITLE
Invoke sagemaker-spark SDK deployed models outside of Spark

### ISSUE THREAD

**Issue by robperc** [aws/sagemaker-spark#33]
Is it possible to invoke models trained and deployed using the sagemaker-spark SDK outside of Spark- for example using the Boto3 SageMaker runtime InvokeEndpoint call? I'm attempting to call my model trained and deployed with the sagemaker-spark SDK using the InvokeEndpoint API call and I get the attached error, but when I load this file into a dataframe in a Spark session and use the model's transform method it works fine. See attached files for my model, test data, and error screenshot.
[model.tar.gz](https://github.com/aws/sagemaker-spark/files/1839093/model.tar.gz)
[mnist.single.test.txt](https://github.com/aws/sagemaker-spark/files/1839097/mnist.single.test.txt)
![error](https://user-images.githubusercontent.com/9399522/37795461-c01c3500-2dd1-11e8-9107-cdc6a7a55988.png)

If this is indeed possible I think an example in the documentation would be very instructive and helpful.

**Comment by robperc** [comment]
Here's a zip containing the .ipynbs used to train and deploy the model on Spark then attempt to invoke using Boto3 on Python3
[notebooks.zip](https://github.com/aws/sagemaker-spark/files/1843236/notebooks.zip)

**Comment by robperc** [comment]
Here's a zip containing the .ipynbs used to train and deploy the model on Spark then attempt to invoke using Boto3 on Python3
[notebooks.zip](https://github.com/aws/sagemaker-spark/files/1843242/notebooks.zip)


*** Re-uploaded so .zip included mnist.single.test data file

**Comment by ChoiByungWook** [comment]
Hello,

Thank you for using SageMaker.

From my understanding, you are attempting to invoke an endpoint for a XGBoost model that was trained using the SparkSDK. As long as your inference data is serialized to the correct format, then you should be able to invoke the endpoint regardless of the SDK.

For details on serialization, please refer to the serializer defined in each corresponding estimator, https://github.com/aws/sagemaker-spark/blob/master/sagemaker-spark-sdk/src/main/scala/com/amazonaws/services/sagemaker/sparksdk/algorithms/XGBoostSageMakerEstimator.scala#L479. 

For XGBoost, the default serializer is for LIBSVM, https://github.com/aws/sagemaker-spark/blob/master/sagemaker-spark-sdk/src/main/scala/com/amazonaws/services/sagemaker/sparksdk/transformation/serializers/LibSVMRequestRowSerializer.scala.

This would require you to serialize the data correctly.

Thanks for reporting this, I'll make sure to update the docs.

Please let me know if there is anything that I can clarify.

**Comment by robperc** [comment]
Hrmm, maybe I'm serializing the test data incorrectly. I'll double check my work and try again. Mind if we leave this open until I validate that this is not a bug?

**Comment by andremoeller** [comment]
Hi @robperc,

The code that normally does the LibSVM serialization is here:

https://github.com/aws/sagemaker-spark/blob/da00fa6c7938341036aa34de3c905f44e51c375c/sagemaker-spark-sdk/src/main/scala/com/amazonaws/services/sagemaker/sparksdk/transformation/serializers/LibSVMRequestRowSerializerUtils.scala#L25-L31

A couple more ideas:
1. Our LibSVM serialization encodes strings in UTF-8. If you're using Python 3, you'll have to do this explicitly: can you try sending your payload with `payload.encode('utf-8')`?
2. You might have to convert your label to a double ('7.0' instead of '7'). I don't know how strict the XGBoost model hosting container is about this. 

Thanks!

**Comment by robperc** [comment]
I tested both cases (int and double labels) against my deployed model in PySpark. You're right that the model does expect a double as can be seen below. Good callout.

<img width="903" alt="working" src="https://user-images.githubusercontent.com/9399522/37940028-4c737c24-311b-11e8-9452-caecc0c6d2f8.png">

When I try to invoke it with an integer label it doesn't work

<img width="1080" alt="notworking" src="https://user-images.githubusercontent.com/9399522/37940083-917c1a10-311b-11e8-839b-1a53acf52ff5.png">

However from the boto3 InvokeEndpoint call I'm still getting the same 400 error. I've tried encoding the payload as utf-8 but it didn't seem to work. Below my code as it stands now. Could it be that the model only accepts a specific content type?

`file_name = 'mnist.single.test'

with open(file_name, 'r') as f:
    payload = f.read()
    
payload = payload.encode('utf-8')

response = runtime_client.invoke_endpoint(EndpointName='my_endpoint', 
                                   ContentType='text/x-libsvm', 
                                   Body=payload)`

**Comment by andremoeller** [comment]
The content type 'text/x-libsvm' seems to be correct. It's what the [XGBoost docs say it can accept](https://docs.aws.amazon.com/sagemaker/latest/dg/xgboost.html), and it's what the Scala SDK uses.

I'm not certain what's going wrong, but a few more ideas:

1. Your LibSVM is dense -- the LibSVM serializer in our SDK ignores zero values. You might want to try either sending just the non-zero values (which I've attached below).

My suspicion is that the model infers the feature dimension from the training data, but that the training data doesn't have any non-zero values for the later indices (784th, 783rd, 782nd, e.g.), and that the model doesn't know how to handle such indices, and so inference fails in the endpoint.

2. Your endpoint logs might say something potentially useful -- if you haven't looked at them already, you can find them in CloudWatch under /aws/sagemaker/Endpoints/[EndpointName] [link to AWS docs on CloudWatch](https://docs.aws.amazon.com/sagemaker/latest/dg/logging-cloudwatch.html).


```
7.0 203:0.328125 204:0.72265625 205:0.62109375 206:0.58984375 207:0.234375 208:0.140625 231:0.8671875 232:0.9921875 233:0.9921875 234:0.9921875 235:0.9921875 236:0.94140625 237:0.7734375 238:0.7734375 239:0.7734375 240:0.7734375 241:0.7734375 242:0.7734375 243:0.7734375 244:0.7734375 245:0.6640625 246:0.203125 259:0.26171875 260:0.4453125 261:0.28125 262:0.4453125 263:0.63671875 264:0.88671875 265:0.9921875 266:0.87890625 267:0.9921875 268:0.9921875 269:0.9921875 270:0.9765625 271:0.89453125 272:0.9921875 273:0.9921875 274:0.546875 292:0.06640625 293:0.2578125 294:0.0546875 295:0.26171875 296:0.26171875 297:0.26171875 298:0.23046875 299:0.08203125 300:0.921875 301:0.9921875 302:0.4140625 327:0.32421875 328:0.98828125 329:0.81640625 330:0.0703125 354:0.0859375 355:0.91015625 356:0.99609375 357:0.32421875 382:0.50390625 383:0.9921875 384:0.9296875 385:0.171875 409:0.23046875 410:0.97265625 411:0.9921875 412:0.2421875 437:0.51953125 438:0.9921875 439:0.73046875 440:0.01953125 464:0.03515625 465:0.80078125 466:0.96875 467:0.2265625 492:0.4921875 493:0.9921875 494:0.7109375 519:0.29296875 520:0.98046875 521:0.9375 522:0.22265625 546:0.07421875 547:0.86328125 548:0.9921875 549:0.6484375 573:0.01171875 574:0.79296875 575:0.9921875 576:0.85546875 577:0.13671875 601:0.1484375 602:0.9921875 603:0.9921875 604:0.30078125 628:0.12109375 629:0.875 630:0.9921875 631:0.44921875 632:0.00390625 656:0.51953125 657:0.9921875 658:0.9921875 659:0.203125 683:0.23828125 684:0.9453125 685:0.9921875 686:0.9921875 687:0.203125 711:0.47265625 712:0.9921875 713:0.9921875 714:0.85546875 715:0.15625 739:0.47265625 740:0.9921875 741:0.80859375 742:0.0703125
```
Thanks!

**Comment by robperc** [comment]
It works!!!
<img width="726" alt="working" src="https://user-images.githubusercontent.com/9399522/37941221-fe0c793a-3121-11e8-84c6-a279b6d7d796.png">

From poking around trying to break it it looks like when I add more than one feature of `index: 0.0` the 400 error pops back up. Not sure why but I guess this just means I'll need to add data prep code to handle removal of sparse entries before passing the data along.

Thanks a ton for all the help! Closing this now. I'll take my notebooks I created while troubleshooting this and [add them as examples here in a bit](https://github.com/robperc/sagemaker-spark/tree/master/examples/notebooks/Jupyter). Keep an eye out for my PR!
